File: src/checkpoint-data/src/main/scala/CheckpointData.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark

import org.apache.spark.ml.param._
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.util.{DefaultParamsReadable, Identifiable}
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.types._
import org.apache.spark.storage._

trait CheckpointDataParams extends MMLParams {

  // Determines the storage level: MEMORY_ONLY or MEMORY_AND_DISK
  val diskIncluded: BooleanParam = BooleanParam(this, "diskIncluded", "Persist to disk as well as memory", false)
  final def getDiskIncluded: Boolean = $(diskIncluded)
  def setDiskIncluded(value: Boolean): this.type = set(diskIncluded, value)

  // Enables reverse operation to free up memory
  val removeCheckpoint: BooleanParam = BooleanParam(this, "removeCheckpoint", "Unpersist a cached dataset", false)
  final def getRemoveCheckpoint: Boolean = $(removeCheckpoint)
  def setRemoveCheckpoint(value: Boolean): this.type = set(removeCheckpoint, value)

  protected def validateAndTransformSchema(schema: StructType): StructType = {
    schema
  }

}

class CheckpointData(override val uid: String) extends Transformer with CheckpointDataParams {

  def this() = this(Identifiable.randomUID("CheckpointData"))

  override def transform(dataset: Dataset[_]): DataFrame = {
    if ($(removeCheckpoint)) {
      CheckpointData.clearCache(dataset, false)
    } else {
      CheckpointData.cache(dataset, $(diskIncluded), false)
    }
  }

  def transformSchema(schema: StructType): StructType = {
    validateAndTransformSchema(schema)
  }

  def copy(extra: ParamMap): CheckpointData = defaultCopy(extra)

}

object CheckpointData extends DefaultParamsReadable[CheckpointData]{

  def clearCache(ds: Dataset[_], blocking: Boolean): DataFrame = {
    ds.unpersist(blocking)
    ds.toDF
  }

  def cache(ds: Dataset[_], disk: Boolean, serialized: Boolean): DataFrame = {
    ds.persist(if (disk && serialized) StorageLevel.MEMORY_AND_DISK_SER
               else if (serialized)    StorageLevel.MEMORY_ONLY_SER
               else if (disk)          StorageLevel.MEMORY_AND_DISK
               else                    StorageLevel.MEMORY_ONLY)
    ds.toDF
  }

  def persistToHive(ds: Dataset[_], dbName: String, tableName: String): DataFrame = {
    ds.write.mode("overwrite").saveAsTable(dbName + "." + tableName)
    ds.toDF
  }

}

File: src/src/main/resources/mmlspark/CheckpointData.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class CheckpointData(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """
    ``CheckpointData`` persists data to disk as well as memory.

    Storage level is MEMORY_AND_DISK if true, else MEMORY_ONLY.
    The default is false (MEMORY_ONLY).

    Use the removeCheckpoint parameter to reverse the cache operation.

    Args:

        diskIncluded (bool): Persist to disk as well as memory (default: false)
        removeCheckpoint (bool): Unpersist a cached dataset (default: false)
    """

    @keyword_only
    def __init__(self, diskIncluded=False, removeCheckpoint=False):
        super(CheckpointData, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.CheckpointData")
        self.diskIncluded = Param(self, "diskIncluded", "diskIncluded: Persist to disk as well as memory (default: false)")
        self._setDefault(diskIncluded=False)
        self.removeCheckpoint = Param(self, "removeCheckpoint", "removeCheckpoint: Unpersist a cached dataset (default: false)")
        self._setDefault(removeCheckpoint=False)
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, diskIncluded=False, removeCheckpoint=False):
        """
        Set the (keyword only) parameters

        Args:

            diskIncluded (bool): Persist to disk as well as memory (default: false)
            removeCheckpoint (bool): Unpersist a cached dataset (default: false)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setDiskIncluded(self, value):
        """

        Args:

            diskIncluded (bool): Persist to disk as well as memory (default: false)

        """
        self._set(diskIncluded=value)
        return self


    def getDiskIncluded(self):
        """

        Returns:

            bool: Persist to disk as well as memory (default: false)
        """
        return self.getOrDefault(self.diskIncluded)


    def setRemoveCheckpoint(self, value):
        """

        Args:

            removeCheckpoint (bool): Unpersist a cached dataset (default: false)

        """
        self._set(removeCheckpoint=value)
        return self


    def getRemoveCheckpoint(self):
        """

        Returns:

            bool: Unpersist a cached dataset (default: false)
        """
        return self.getOrDefault(self.removeCheckpoint)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.CheckpointData"

    @staticmethod
    def _from_java(java_stage):
        module_name=CheckpointData.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".CheckpointData"
        return from_java(java_stage, module_name)
